### ISSUE TITLE
Scalability Improvement

### ISSUE THREAD

**Issue by ccoffrin** [lanl-ansi/QuantumAnnealing.jl#14]
I am getting a `ERROR: StackOverflowError:` when building a system with 9 or more qubits.  Tested on Julia v1.7.0.

MWE:
```
using QuantumAnnealing
n = 9
@time ρ = simulate(Dict((i,) => 1.0 for i in 1:n), 1.0, AS_CIRCULAR)
```

**Comment by zmorrell** [comment]
@ccoffrin That is a surprisingly low number of qubits.  I just tried to replicate on my own machine and was able to get up to 12 qubits successfully, with the process being killed (likely due to lack of memory) at 13 qubits.  The following is my output.  I am running Julia 1.7.1 on a system with 8 GB of RAM.  I do agree that we should ideally improve the scalability, I am just surprised that we are having such a large difference in number of qubits and error modes.

```
n = 12

iter |  steps  |    max(Δ)    |    mean(Δ)   |
   1 |       4 | 5.372611e-03 | 3.621762e-05 |
   2 |       8 | 4.046612e-05 | 4.065840e-07 |

converged
   iterations........: 2
   simulation steps..: 8
   maximum difference: 4.046612e-05 <= 1.000000e-04
   mean difference...: 4.065840e-07 <= 1.000000e-06
   runtime (seconds).: 531.093272

534.307696 seconds (22.98 M allocations: 42.642 GiB, 4.51% gc time, 2.05% compilation time)
n = 13

iter |  steps  |    max(Δ)    |    mean(Δ)   |
Killed
```

**Comment by ccoffrin** [comment]
Thanks for double checking this. I confirmed the issue was in Julia, I updated to 1.7.2 on OS X and it resolved the issue.  The code easily scaled to 12 qubits on a MacBook Pro with 32GB of memory.  At 13 qubits CPU time is currently the primary bottle neck.

I started using this for testing to see how far I can go,
```
for n in 1:20
    println("qubits: $(n)")
    @time ρ = simulate(Dict((i,) => 1.0 for i in 1:n), 1.0, AS_CIRCULAR);
end
```

**Comment by ccoffrin** [comment]
As the stack overflow issue is resolved I am closing this in favor of #5.

For reference, here are the basic runtime and memory requirements,
```
qubits: 2
  0.003559 seconds (15.56 k allocations: 1.059 MiB)
...
qubits: 4
  0.004150 seconds (17.19 k allocations: 4.588 MiB)
...
qubits: 8
  0.333331 seconds (13.69 k allocations: 340.731 MiB, 11.26% gc time)
qubits: 9
  1.709307 seconds (15.37 k allocations: 1.300 GiB, 17.53% gc time)
qubits: 10
  7.513668 seconds (17.07 k allocations: 5.118 GiB, 9.91% gc time)
qubits: 11
 25.136623 seconds (12.62 k allocations: 10.044 GiB, 5.70% gc time)
qubits: 12
210.228679 seconds (14.55 k allocations: 41.478 GiB, 3.48% gc time)
qubits: 13
1709.890691 seconds (16.41 k allocations: 165.503 GiB, 1.07% gc time)
```

Basic memory requirements (without matrix exponent),
```
qubits: 9
  0.017954 seconds (5.12 k allocations: 34.486 MiB)
qubits: 10
  0.074035 seconds (5.96 k allocations: 122.052 MiB, 11.28% gc time)
qubits: 11
  0.240921 seconds (7.25 k allocations: 456.728 MiB, 7.82% gc time)
qubits: 12
  1.231461 seconds (8.51 k allocations: 1.710 GiB, 29.96% gc time)
qubits: 13
  4.123376 seconds (9.75 k allocations: 6.689 GiB, 16.33% gc time)
qubits: 14
 20.451149 seconds (11.10 k allocations: 26.401 GiB, 6.30% gc time)
qubits: 15
332.886369 seconds (12.53 k allocations: 104.888 GiB, 1.87% gc time)
```
